File: CSharp-Part-2/6. Strings and Text Processing/StringsAndTextProcessing/ParseTags/ParseTags.cs
/*
 * You are given a text. Write a program that changes the text in all regions surrounded by the tags <upcase> and </upcase> to upper-case.
 * The tags cannot be nested.
 */

using System;
using System.Text;

class ParseTags
{
    static void Main()
    {
        string inputStr = "We are living in a <upcase>yellow submarine</upcase>. We don't have <upcase>anything</upcase> else.";

        bool isTag = false;
        bool isInUpcase = false;
        StringBuilder finalStr = new StringBuilder();
        for (int i = 0; i < inputStr.Length; i++)
        {
            // Match an opening tag
            if (inputStr[i] == '<' && inputStr[i + 1] != '/' && i != inputStr.Length - 1)
            {
                isTag = true;
                continue;
            }
            // Match a closing tag
            if (inputStr[i] == '<' && inputStr[i + 1] == '/')
            {
                isInUpcase = false;
                isTag = true;
            }
            if (i > 0 && inputStr[i - 1] == '>')
            {
                // Only capitalize the next sequence if the tag isn't closing
                if (inputStr[i - 8] != '/')
                {
                    isInUpcase = true;
                }
                isTag = false;
            }
            if (isInUpcase)
            {
                finalStr.Append(Char.ToUpper(inputStr[i]));
            }
            else if (!isTag)
            {
                finalStr.Append(inputStr[i]);
            }
        }
        Console.WriteLine(finalStr);
    }
}

File: JavaScript-Fundamentals/11. Strings/parse-tags.js
/*
	You are given a text. Write a function that changes the text in all regions:

		<upcase>text</upcase> to uppercase.
		<lowcase>text</lowcase> to lowercase
		<mixcase>text</mixcase> to mix casing (random)
*/

var text = "We are <mixcase>living</mixcase> in a <upcase>yellow submarine</upcase>. We <mixcase>don't</mixcase> have <lowcase>anything</lowcase> else.";

function parseTags(text) {
    var i,
        shift,
        currentTags = [],
        result = [];

    for (i = 0, len = text.length; i < len; i += 1) {
        if (text[i] === '<') {
            switch (text[i + 1]) {
                case 'u':
                    currentTags.push('u');
                    shift = text.indexOf('>', i);
                    i = shift + 1;
                    break;
                case 'l':
                    currentTags.push('l');
                    shift = text.indexOf('>', i);
                    i = shift + 1;
                    break;
                case 'm':
                    currentTags.push('m');
                    shift = text.indexOf('>', i);
                    i = shift + 1;
                    break;
                case '/':
                    currentTags.pop();
                    shift = text.indexOf('>', i);
                    i = shift + 1;
                    break;
            }
        }

        switch (currentTags[currentTags.length - 1]) {
            case undefined:
                result.push(text[i]);
                break;
            case 'u':
                result.push(text[i].toUpperCase());
                break;
            case 'l':
                result.push(text[i].toLowerCase());
                break;
            case 'm':
                if (Math.random() < 0.5) {
                    result.push(text[i].toLowerCase());
                    break;
                } else {
                    result.push(text[i].toUpperCase());
                    break;
                }
        }
    }

    return result.join('');
}

console.log(parseTags(text));